### DISCUSSION TITLE
How can I know the exact meaning of a AST node.

### DISCUSSION THREAD

**Discussion by yszc** [JSQLParser/JSqlParser#2004] [Q&A] [answered]
I'm developing a custom SQL processor using the visitor pattern.
How can I know the exact meaning of a AST class, such as `AllColumns`, `AllTableColumns`, `AllValue`.
What's the origin expression it related in a SQL statement.

**Comment by manticore-projects** [comment]
Greetings.

I understand what you mean, this kind of documentation is missing:

`AllColumns` refers to the `*` in `SELECT * FROM ...`
`AllTableColumns` with `tablename='t'` refers to to `t.*` in `SELECT t.* FROM ...`

A strong volunteer would need to write that as Java-Doc  into each AST class.
In the meantime, your best bet was to go through those AST classes and study their `toString()` method in order to understand what this class is holding.

**Reply by yszc** [reply]
Thanks, but I can't use `toString()`.

The situation is, I need to process all kinds of nodes extended by `ASTNodeAccessImpl`, so there are some grammer beyond my knowledge, so I need to get SQL snippet directly from the class (have no object).

Now I have a method which is searching for the class name in `JSqlParserCC.jjt` and analyze the context.  Is there any more efficiency method?

**Reply by manticore-projects** [reply, marked as answer]
I am sorry to tell you that exactly this kind of documentation is missing, because nobody has the time and resources to write it.
Of course you are most welcome to submit PRs with JavaDoc when you are on this job now.

Btw, IntelliJ has an excellent JavaCC plugin which makes it easy to find the production for each AST class.
Also, you could parse the grammar with javaCC itself to get the BNF of each production and last but not least https://manticore-projects.com/JSQLParser/syntax_snapshot.html will show you the Grammar and has a searchable side bar.


Example for `AllTableColumns`: https://manticore-projects.com/JSQLParser/syntax_snapshot.html#alltablecolumns

**Reply by yszc** [reply]
Thank you, I wasn't expecting a documentation-based answer.  The Syntax Snapshot is very useful.  😀

**Reply by manticore-projects** [reply]
You are welcome and I am glad to be of assistance. If anything else comes up, please don't hesitate to ask. Cheers!

**Comment by manticore-projects** [comment]
Additionally you can use [JSQLFormatter online](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20M4UwNiDGAuAE0DoBUsBmAnA9gW1gEwFcBDMIA)

to get this kind of AST illustration:
```
SQL Text
 └─Statements: statement.select.PlainSelect
    ├─selectItems: statement.select.SelectItem
    │  └─expression: statement.select.AllTableColumns
    │     └─Table: t
    └─Table: dual
```
